veracruz-client/src/veracruz_client.rs

```diff
@@ -150,66 +220,56 @@ impl VeracruzClient {
+        let policy_ciphersuite = veracruz_utils::lookup_ciphersuite_mbedtls(
+            policy.ciphersuite().as_str(),
+        )
+        .ok_or_else(|| {
+            VeracruzClientError::TLSInvalidCiphersuiteError(policy.ciphersuite().to_string())
+        })?;
+        let cipher_suites: Vec<i32> = vec![policy_ciphersuite.into(), 0];
+        config.set_ciphersuites(Arc::new(cipher_suites));
+        let entropy = Arc::new(mbedtls::rng::OsEntropy::new());
+        let rng = Arc::new(mbedtls::rng::CtrDrbg::new(entropy, None)?);
+        config.set_rng(rng);
+        config.set_ca_list(Arc::new(proxy_service_cert), None);
+        config.push_cert(Arc::new(client_cert), Arc::new(client_priv_key))?;
+        let mut ctx = mbedtls::ssl::Context::new(Arc::new(config));
+        let remote_session_id = Arc::new(Mutex::new(Some(0)));
+        let conn = InsecureConnection {
+            read_buffer: vec![],
+            veracruz_server_url: policy.veracruz_server_url().to_string(),
+            remote_session_id: Arc::clone(&remote_session_id),
         };
-        let mut root_store = rustls::RootCertStore::empty();
-        root_store
-            .add(&rustls::Certificate(proxy_service_cert))
-            .map_err(VeracruzClientError::WebpkiError)?;
-
-        let policy_ciphersuite = veracruz_utils::lookup_ciphersuite(policy_ciphersuite_string)
-            .ok_or_else(|| {
-                VeracruzClientError::TLSInvalidCiphersuiteError(
-                    policy_ciphersuite_string.to_string(),
-                )
-            })?;
-
-        let client_config = rustls::ClientConfig::builder()
-            .with_cipher_suites(&[policy_ciphersuite])
-            .with_safe_default_kx_groups()
-            .with_protocol_versions(&[&rustls::version::TLS12])?
-            .with_root_certificates(root_store)
-            .with_single_cert([client_cert].to_vec(), client_priv_key)?;
-
-        let enclave_name_as_server = rustls::ServerName::try_from(enclave_name)
-            .map_err(VeracruzClientError::InvalidDnsNameError)?;
-        let connection =
-            ClientConnection::new(std::sync::Arc::new(client_config), enclave_name_as_server)?;
-        let client_cert_text = VeracruzClient::read_all_bytes_in_file(&client_cert_filename)?;
-        let mut client_cert_raw = from_utf8(client_cert_text.as_slice())?.to_string();
-        // erase some '\n' to match the format in policy file.
-        client_cert_raw.retain(|c| c != '\n');
-        let client_cert_string = client_cert_raw
-            .replace(
-                "-----BEGIN CERTIFICATE-----",
-                "-----BEGIN CERTIFICATE-----\n",
-            )
-            .replace("-----END CERTIFICATE-----", "\n-----END CERTIFICATE-----");
+        ctx.establish(conn, None)?;
 
         Ok(VeracruzClient {
-            tls_connection: connection,
-            remote_session_id: None,
+            tls_connection: ctx,
```

Comment:
Would it make sense to rename `tls_connection` to `tls_context`?

Reply:
Probably. I'll do that.